lib/services/nodes-api-service.js

```diff
@@ -30,64 +32,136 @@ function nodeApiServiceFactory(
-            if (index === -1 || !_.has(enclNode.relations[index], 'targets')) {
-                return;
-            }
+        var index = _.findIndex(node.relations, { relationType: type });
+        if (index === -1 || !_.has(node.relations[index], 'targets')) {
+            return Promise.resolve();
+        }
+
+        // Remove target node id in relation field
+        if (_.indexOf(node.relations[index].targets, targetId) !== -1) {
+            _.pull(node.relations[index].targets, targetId);
+        }
+
+        // Remove the type of relation if no targets in it
+        if (node.relations[index].targets.length === 0) {
+            _.pull(node.relations, node.relations[index]);
+        }
+
+        return waterline.nodes.updateByIdentifier(node.id,
+                                                  {relations: node.relations});
+    };
+
+    /**
+     * Handle target nodes. Update relation type field and remove the target
+     * node if needed.
+     * @param  {Object}     targetNodos array of target nodes
+     * @param  {Object}     type relation type in the removing node
+     * @param  {Object}     nodeId removing node id
+     * @return {Promise}
+     */
+    NodeApiService.prototype._handleTargetNodes = function(targetNodes, type, nodeId) {
+        var self = this;
+        var targetType = '';
+        var relationClass = null;
+        var needDel = false;
+
+        if (Constants.NodeRelations.hasOwnProperty(type)) {
+            targetType = Constants.NodeRelations[type].mapping;
+            relationClass = Constants.NodeRelations[type].relationClass;
+        }
+
+        if ((relationClass === 'component') &&
+            (type.indexOf('By') === -1)) {
+            // Its components need to be deleted
+            needDel = true;
+        }
 
-            if (_.indexOf(enclNode.relations[index].targets, node.id) !== -1) {
-                _.pull(enclNode.relations[index].targets, node.id);
+        return Promise.resolve()
+        .then(function () {
+            if (needDel === true) {
+                // Throw error when all target nodes need to be deleted
+                // but at least one of them have active workflow
+                return Promise.map(targetNodes, function(targetNode) {
+                    return self._delValidityCheck(targetNode.id);
+                });
             }
-            // If Enclosure node doesn't have enclosed nodes
-            // remove this enclosure node, else remove relationships
-            if (enclNode.relations[index].targets.length === 0) {
-                return waterline.nodes.destroy({ id: enclNode.id });
-            } else {
-                return waterline.nodes.updateByIdentifier(
-                    enclNode.id, { relations : enclNode.relations });
+        })
+        .then(function () {
+            return Promise.map(targetNodes, function (targetNode) {
+                // Delete target nodes if it is reuiqred
+                if (needDel === true) {
```

Comment:
Do you want to remove the node and then the relation, or the relation and then the node? Or does it matter?